Patchable Installer

papeh opened the pull request:
- Add GenericInstaller as a submodule
- Connect installer target of build script to build a new base installer
- Sign automatically as able
- Remove dead build targets

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="25" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/flexbridge/170)
<!-- Reviewable:end -->

ermshiperete commented:
I didn't do a careful review, but before I forget just a few comments from glancing over it.

---

Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[.gitignore, line 3 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFnv3jnGYeJh_k6V6p:-KyFnv3jnGYeJh_k6V6q:b-xf51fk) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/.gitignore#L3)):*
> ```
> src/_ReSharper.LiftBridge/
> bin/
> BuildDir
> ```

If this is a directory then it should end with a slash

---

*[BuildBase.bat, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFnMeSLgZi8mDYdSf-:-KyFnMeSLgZi8mDYdSf0:bl5pbrx) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildBase.bat#L1)):*
> ```Batchfile
> @echo off
> ```

IMO this file would be better located in the `build` subdirectory

---

*[BuildBase.bat, line 10 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFn_ijs8PFupPs1zYq:-KyFn_ijs8PFupPs1zYr:b-7y2be6) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildBase.bat#L10)):*
> ```Batchfile
> REM Run the next target only if the previous target succeeded
> (
> 	MSBuild Sample.targets /t:DownloadBuildTasks
> ```

I don't see a  `Sample.targets` file

---

*[BuildPatch.bat, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFn_1MDBSu3aA8nSYm:-KyFn_1Nd2IkzSyrOr1x:bl5pbrx) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildPatch.bat#L1)):*
> ```Batchfile
> @echo off
> ```

IMO this file would be better located in the `build` subdirectory

---

*[DistFiles/License.htm, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFoMxk6NjCNLxp3j3_:-KyFoMxlv7hEqG5bBhVW:bb8v6nz) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/DistFiles/License.htm#L1)):*
> ```HTML
> <html>
> ```

Could this be markdown or at least plain HTML (without the gibberish that Word added)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-:-KyFo01ov-RDlqOKPcGI:buk7szr)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[BuildBase.bat, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFnMeSLgZi8mDYdSf-:-KyLvtuFKTlfaOUBkm9i:b-wzk67k) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildBase.bat#L1)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

IMO this file would be better located in the `build` subdirectory
</blockquote></details>

As it happens, this file doesn't need to be in this repository at all.  It was copied from the genericinstaller-sample project but never used here.

---

*[DistFiles/License.htm, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFoMxk6NjCNLxp3j3_:-KyLwCLz20QpKya1ayjv:b-2gh6ik) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/DistFiles/License.htm#L1)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Could this be markdown or at least plain HTML (without the gibberish that Word added)?
</blockquote></details>

I cleaned it up a bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

papeh commented:
TODO: install FLEx Bridge's  registry keys where FLEx expects them.

---

Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 3 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFnv3jnGYeJh_k6V6p:-KyvRQvYDnCTiGzTUz7z:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/.gitignore#L3)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

If this is a directory then it should end with a slash
</blockquote></details>

Done.

---

*[BuildBase.bat, line 10 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFn_ijs8PFupPs1zYq:-KyvRjApvPZjFGNRwebk:b-jw74f) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildBase.bat#L10)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

I don't see a  `Sample.targets` file
</blockquote></details>

This file was copied from the genericinstaller-sample project but never used here. It has been removed.

---

*[BuildPatch.bat, line 1 at r1](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KyFn_1MDBSu3aA8nSYm:-KyvRhppWZEtEXlKmVjH:b-h4q4sn) ([raw file](https://github.com/sillsdev/flexbridge/blob/5b32b1275f257abb4d1728e3944c9972c46d0aaa/BuildPatch.bat#L1)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

IMO this file would be better located in the `build` subdirectory
</blockquote></details>

This file was copied from the genericinstaller-sample project but never used here. It has been removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-:-KyvRss6PoycncZirXMe:b9dxrcd)*
<!-- Sent from Reviewable.io -->

marksvc commented:
Thanks for all your work on this. See comments.

---

Reviewed 6 of 13 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[ReadMe.md, line 99 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZfcasqGdYuivUZnKN:-KzZfcasqGdYuivUZnKO:b-e21k95) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/ReadMe.md#L99)):*
> buid

Typo "build".

"rerunner" should probably have a hyphen in it.

---

*[build/FLExBridge.proj, line 141 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZgrYYPSpWT6RmxzJ5:-KzZgrYYPSpWT6RmxzJ6:b-55pntk) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/FLExBridge.proj#L141)):*
> ```XML
> 		</ReadLinesFromFile>
> 		<CreateProperty Condition="'$(BuildCounter)' != ''" Value="$(Version).$(BuildCounter)">
> 			<Output TaskParameter="Value" PropertyName="Version"/>
> ```

Hmm, so is this overwriting the existing value of the 'Version' property with 'Version.BuildCounter' ?

---

*[build/WixPatchableInstaller.targets, line 21 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZlqWWn3VFR2e2-5Cw:-KzZlqWXDrMSnx9v3x6C:bjb2iv2) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L21)):*
> ```XML
> 	<!-- $(ApplicationName) 1.0.0.X ($(ApplicationName) build)-->
> 	<PropertyGroup>
> 		<ProductIdGuid>57942FAA-4E7F-447F-8323-795CEF93FBA1</ProductIdGuid>
> ```

This seems like something that should be generated, not edited in this file. Or is this placed here by a process that automatically generated the GUID and stuck it in here?

---

*[build/WixPatchableInstaller.targets, line 26 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZlzB4l6bXrgPRWaVp:-KzZlzB4l6bXrgPRWaVq:bxe7bu) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L26)):*
> ```XML
> 	<!-- $(ApplicationName) 1.0.1.X ($(ApplicationName) build)-->
> 	<!-- <PropertyGroup>
> 	<! -	<ProductIdGuid>E4789AE1-EE7A-4488-88E2-BD3B633DA1E0</ProductIdGuid>
> ```

What's with this commented out section? Or is this an example of what will need to happen for a future version number? This strikes me as something that will quickly become a dusty artifact of the code, and so maybe this section should be removed now/before then?

---

*[build/WixPatchableInstaller.targets, line 34 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZmS1LR_a-m2na3GJP:-KzZmS1MwAJ-1wnKpPl1:bzaon37) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L34)):*
> ```XML
> 	<!-- ### one another because they share a common upgrade code.                                              ### -->
> 	<!-- ### Don't even THINK about changing the Upgrade Code. Ever.                                            ### -->
> 	<!-- ########################################################################################################## -->
> ```

These
########################################################################################################## 
lines make my eyes a bit sore. It's not a big deal that they're here, but if you don't like them either, you could consider whether they should stay.

---

*[build/WixPatchableInstaller.targets, line 70 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZoZcCUNGwn4iufTpC:-KzZoZcCUNGwn4iufTpD:b-gmvl0n) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L70)):*
> ```XML
> 			<VersionSeg4 Condition="'$(VersionSeg4)'==''">1</VersionSeg4> <!-- 1 is the base build number. It may be a NAnt thing... -->
> 
> 			<!-- REVIEW (Hasso) 2017.09: what do we call each segment again? Major.Minor.(Build|Patch).(Revision|Build) -->
> ```

If helpful, https://semver.org/ defines the 3rd element as Patch. I'm not sure how closely FB versioning follows that model though.

---

*[build/WixPatchableInstaller.targets, line 73 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZnZA34r2D9LIt6egv:-KzZnZA34r2D9LIt6egw:b-3plg5d) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L73)):*
> ```XML
> 			<MajorVersion>$(VersionSeg1).$(VersionSeg2)</MajorVersion>
> 			<MinorVersion>$(MajorVersion).$(VersionSeg3)</MinorVersion>
> 			<Revision>$(MinorVersion).$(VersionSeg4)</Revision>
> ```

I would have more likely expected this to say

```
<MajorVersion>$(VersionSeg1)</MajorVersion>
<MinorVersion>$(VersionSeg1).$(VersionSeg2)</MinorVersion>
<Whatever>$(VersionSeg1).$(VersionSeg2).$(VersionSeg3)</Whatever>
<Revision>$(VersionSeg1).$(VersionSeg2).$(VersionSeg3).$(VersionSeg4)</Revision>
```

As it is, won't this be defining MajorVersion to include the first 2 elements of the version code?

---

*[build/WixPatchableInstaller.targets, line 100 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZq0OtHXpHYZk4-Dk2:-KzZq0OuJSF7Rf9eclnn:bgdp5l2) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L100)):*
> ```XML
> 	<!-- ########################################################################################################## -->
> 	<!-- ### Build Targets                                                                                      ### -->
> 	<!-- ### TODO (Hasso) 2107.10: assimilate with main targets                                                 ### -->
> ```

Does this imply optimism that WIX will work on a Windows from the 22nd century?

---

*[build/WixPatchableInstaller.targets, line 109 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZqWyVBQpaNABcxFgr:-KzZqWyVBQpaNABcxFgs:b-tpamic) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L109)):*
> Configuration=release;Platform=x86

I'm used to seeing 'Release' with a capital 'R'. Does that not matter here?

I'm surprised the Platform is x86 instead of something like AnyCPU. ?

---

*[build/WixPatchableInstaller.targets, line 209 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZy_Ib-EnxzaUdH36e:-KzZy_Ib-EnxzaUdH36f:b23qpmr) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L209)):*
> ```XML
> 			<MsiFile>$(SafeApplicationName)_$(Revision).msi</MsiFile>
> 			<BaseBuildDir>$(RootDir)/src/WiXInstaller/BaseInstallerBuild</BaseBuildDir>
> 			<BaseBuildArgs><![CDATA["$(ApplicationName)" $(SafeApplicationName) $(Revision) $(ProductIdGuid) $(UpgradeCodeGuid) "$(AppBuildDir)/$(BinDirSuffix)" "$(AppBuildDir)/$(DataDirSuffix)" $(CopyrightYear) "$(Manufacturer)" "$(CertPath)" "$(CertPass)"]]></BaseBuildArgs>
> ```

Why is this in a CDATA?

---

*[DistFiles/about.htm, line 35 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZzxlW2Ovxwh3raNiY:-KzZzxlW2Ovxwh3raNiZ:b-n8faou) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/DistFiles/about.htm#L35)):*
> ```HTML
> 					<tr>
> 						<td><strong>Website:</strong></td>
> 						<td><a href="https://github.com/sillsdev/libpalaso" target="_blank">FLEx Bridge</a></td>
> ```

The libpalaso page?
Not https://github.com/sillsdev/flexbridge ?

---

*[DistFiles/license.rtf, line 0 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_2Oioz5TT4l_UoRmz:-Kz_2Oioz5TT4l_UoRn-:biah5n6) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/DistFiles/license.rtf#L0)):*
What's in license.rtf? The same content as License.htm? Is it helpful to include an .rtf version of the license?

I see it's referred to in TODO.xml, but maybe that could use the .htm one instead? 

It's not a huge deal. 

---

*[src/Installer/TODO.xml, line 4 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_5lQ2CWArZQYNfF6X:-Kz_5lQ2CWArZQYNfF6Y:b-b5e1gl) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L4)):*
> ```XML
> <WixVariable Id="WixUIBannerBmp" Value="BannerWithSILLogo.bmp" />
> <WixVariable Id="WixUIDialogBmp" Value="DialogBGWithSILLogo.bmp" />
> ```

Okay after reviewing this file, I remembered it's called TODO.xml. So does any of this do anything? Is this where you are wanting to store half-finished code?

---

*[src/Installer/TODO.xml, line 10 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_5A3doIFDVLcXIjhe:-Kz_5A3doIFDVLcXIjhf:b4r0ygq) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L10)):*
> ```XML
> <Property Id="MANUFACTURERREGKEY">SIL</Property>
> <!-- The following will set properties A[FW9INSTALLDIR] and B[FW9DEVINSTALLDIR] to the result of two different registry searches.
> If the search for B[FW9DEVINSTALLDIR] was successful, it overrides the value of A[FW9INSTALLDIR] with the value of B[FW9DEVINSTALLDIR].
> ```

Okay but it doesn't actually do this presently, right?

---

*[src/Installer/TODO.xml, line 16 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_3a-6RopzCmUeH_B1:-Kz_3a-6RopzCmUeH_B2:b-99f6lu) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L16)):*
> ```XML
>   <RegistrySearch Id="SearchForFW9" Root="HKLM" Key="SOFTWARE\[MANUFACTURERREGKEY]\FieldWorks\9" Name="RootCodeDir" Type="raw"/>
> </Property>
> <Property Id="FIELDWORKSMINIMUMINSTALLEDVERSION">
> ```

Instead of "FieldWorks Minimum Installed Version", isn't what's cared about here the "FieldWorks Installed Version"? Wouldn't "FieldWorks Minimum Installed Version" imply , for example, that if the user has FW 8.1 and FW 9.1 installed, that the "FieldWorks Minimum Installed Version" is FW 8.1? And that's.. not what we're getting at here, is it?

---

*[src/Installer/TODO.xml, line 34 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_4mLRt4E8hzCvMiMV:-Kz_4mLRt4E8hzCvMiMW:b7orp7g) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L34)):*
> ```XML
> 
> <!-- Use the dev entry, but only if set.
> <SetProperty Id="FW9INSTALLDIR" Before="SetFIELDWORKSMINIMUMINSTALLEDVERSION" Value="[DEVFW9BUILDDIR]">DEVFW9BUILDDIR</SetProperty> -->
> ```

Why is this commented out?

---

*[src/Installer/TODO.xml, line 35 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_5MPNZBO3OpaYeJui:-Kz_5MPNZBO3OpaYeJuj:b7ik6ey) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L35)):*
> ```XML
> <!-- Use the dev entry, but only if set.
> <SetProperty Id="FW9INSTALLDIR" Before="SetFIELDWORKSMINIMUMINSTALLEDVERSION" Value="[DEVFW9BUILDDIR]">DEVFW9BUILDDIR</SetProperty> -->
> <SetProperty Id="FIELDWORKSMINIMUMINSTALLEDVERSION" Before="LaunchConditions" Value="[DEVFIELDWORKSMINIMUMINSTALLEDVERSION]">DEVFIELDWORKSMINIMUMINSTALLEDVERSION</SetProperty>
> ```

I'm pretty confused by what this line is doing. Is `Value=` like saying `DoThisIfThisIsNonZero=` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-:-Kz_8Wyc5n3-LOZflnvt:bmkkjt2)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[ReadMe.md, line 99 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZfcasqGdYuivUZnKN:-Kz_PUeK2HYo0uhDRMZO:b-rvt0mw) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/ReadMe.md#L99)):*
> buid
done
> rerunning
I prefer to assert my prerogative as a native speaker

---

*[build/FLExBridge.proj, line 141 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZgrYYPSpWT6RmxzJ5:-Kz_Q5_YscMMezO-RPuP:bb7sa0c) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/FLExBridge.proj#L141)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Hmm, so is this overwriting the existing value of the 'Version' property with 'Version.BuildCounter' ?
</blockquote></details>

yes.

---

*[build/WixPatchableInstaller.targets, line 21 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZlqWWn3VFR2e2-5Cw:-Kz_QJbgcTvQLk91ixHA:bkdfl6h) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L21)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

This seems like something that should be generated, not edited in this file. Or is this placed here by a process that automatically generated the GUID and stuck it in here?
</blockquote></details>

that'd be nice. I could ask Glenn what the Paratext team does in their release process, or figure something out myself, for a follow-on change.

---

*[build/WixPatchableInstaller.targets, line 26 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZlzB4l6bXrgPRWaVp:-Kz_RSdpzjeSHJa5z04q:brpmqc3) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L26)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

What's with this commented out section? Or is this an example of what will need to happen for a future version number? This strikes me as something that will quickly become a dusty artifact of the code, and so maybe this section should be removed now/before then?
</blockquote></details>

removed.

---

*[build/WixPatchableInstaller.targets, line 34 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZmS1LR_a-m2na3GJP:-Kz_S4s9IE8rd59Isa17:b-68bd6p) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L34)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

These
########################################################################################################## 
lines make my eyes a bit sore. It's not a big deal that they're here, but if you don't like them either, you could consider whether they should stay.
</blockquote></details>

They're prettier in green. How would you feel about`<!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->`?

---

*[build/WixPatchableInstaller.targets, line 70 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZoZcCUNGwn4iufTpC:-Kz_TPCrLuac8eRnlgl1:bpg8a11) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L70)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

If helpful, https://semver.org/ defines the 3rd element as Patch. I'm not sure how closely FB versioning follows that model though.
</blockquote></details>

We can play with how WiX's transformer and patcher agree with that.

---

*[build/WixPatchableInstaller.targets, line 73 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZnZA34r2D9LIt6egv:-Kz_T_AxKE5aKA9XOyZf:b9ffttz) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L73)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

I would have more likely expected this to say

```
<MajorVersion>$(VersionSeg1)</MajorVersion>
<MinorVersion>$(VersionSeg1).$(VersionSeg2)</MinorVersion>
<Whatever>$(VersionSeg1).$(VersionSeg2).$(VersionSeg3)</Whatever>
<Revision>$(VersionSeg1).$(VersionSeg2).$(VersionSeg3).$(VersionSeg4)</Revision>
```

As it is, won't this be defining MajorVersion to include the first 2 elements of the version code?

</blockquote></details>

You're right; maybe I should think on this after Thanksgiving.

---

*[build/WixPatchableInstaller.targets, line 100 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZq0OtHXpHYZk4-Dk2:-Kz_U1J3OSh9YOH7UOr2:b-xudrv2) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L100)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Does this imply optimism that WIX will work on a Windows from the 22nd century?
</blockquote></details>

Actually, I think I'm done with that. I should tell my future self not to worry.

---

*[build/WixPatchableInstaller.targets, line 109 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZqWyVBQpaNABcxFgr:-Kz_UvJ4swQ3wDxW_aH6:bn6lszm) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L109)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

> Configuration=release;Platform=x86

I'm used to seeing 'Release' with a capital 'R'. Does that not matter here?

I'm surprised the Platform is x86 instead of something like AnyCPU. ?
</blockquote></details>

copy-pasted from the sample project.

---

*[build/WixPatchableInstaller.targets, line 209 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZy_Ib-EnxzaUdH36e:-Kz_V-nTJv7trG7ytPqu:bhotoud) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L209)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Why is this in a CDATA?
</blockquote></details>

You're right, it doesn't seem to have any troublesome characters.

---

*[DistFiles/about.htm, line 35 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZzxlW2Ovxwh3raNiY:-Kz_VQdw-xzPEuTwSVL1:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/DistFiles/about.htm#L35)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

The libpalaso page?
Not https://github.com/sillsdev/flexbridge ?
</blockquote></details>

Done.

---

*[DistFiles/license.rtf, line  at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_2Oioz5TT4l_UoRmz:-Kz_VaLtyXgX2oiDkksp:brczqw1) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/DistFiles/license.rtf#L)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

What's in license.rtf? The same content as License.htm? Is it helpful to include an .rtf version of the license?

I see it's referred to in TODO.xml, but maybe that could use the .htm one instead? 

It's not a huge deal. 
</blockquote></details>

I had deleted the license.rtf file and moved its content to License.htm. Perhaps I should update the copyright banners.

---

*[src/Installer/TODO.xml, line 4 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_5lQ2CWArZQYNfF6X:-Kz_VmxOQt8RFC4p1JHW:b-2xtyli) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/src/Installer/TODO.xml#L4)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Okay after reviewing this file, I remembered it's called TODO.xml. So does any of this do anything? Is this where you are wanting to store half-finished code?
</blockquote></details>

These are some things yet TODO before the old installer disappears. The file could use a comment to that effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

*[DistFiles/license.rtf, line  at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-Kz_2Oioz5TT4l_UoRmz:-Kz_XcrcaqY-AWFA_dDa:b-2ijdmj) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/DistFiles/license.rtf#L)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

I had deleted the license.rtf file and moved its content to License.htm. Perhaps I should update the copyright banners.
</blockquote></details>

(The genericinstaller uses License.htm on its "check this box to accept" screen)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 10 of 265 files reviewed at latest revision, 22 unresolved discussions.

---

*[src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs, line 2 at r3](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzdHbxa-w8-gO6l_X2I:-KzdHbxbyf470q-4lwjP:b6yb00o) ([raw file](https://github.com/sillsdev/flexbridge/blob/fcaad64c3900d50fd754abc108b596bdfcf4949b/src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs#L2)):*
> License.htm

It might make more sense to change this to `LICENSE` in the file headers because that exists in the root directory (where many open source projects have a license file). I didn't realize we actually had a `license.rtf`/`license.html` file because I never looked through all the subdirectories...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Review status: 10 of 265 files reviewed at latest revision, 21 unresolved discussions.

---

*[src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs, line 2 at r3](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzdHbxa-w8-gO6l_X2I:-KzdJCjQcsGVqXFG5Ex_:bey5kgh) ([raw file](https://github.com/sillsdev/flexbridge/blob/fcaad64c3900d50fd754abc108b596bdfcf4949b/src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs#L2)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

> License.htm

It might make more sense to change this to `LICENSE` in the file headers because that exists in the root directory (where many open source projects have a license file). I didn't realize we actually had a `license.rtf`/`license.html` file because I never looked through all the subdirectories...
</blockquote></details>

...or remove `(See: License.htm file)` altogether.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: 8 of 266 files reviewed at latest revision, 22 unresolved discussions.

---

*[src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs, line 2 at r3](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzdHbxa-w8-gO6l_X2I:-Kzy6V_zmqwI5QdPAdFw:b-dw60id) ([raw file](https://github.com/sillsdev/flexbridge/blob/fcaad64c3900d50fd754abc108b596bdfcf4949b/src/FLEx-ChorusPlugin/Infrastructure/ActionHandlers/ObtainProjectStrategyFlex.cs#L2)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

...or remove `(See: License.htm file)` altogether.
</blockquote></details>

So there are two files to keep up-to-date. I'll fix that as well as I can in 20 minutes or less.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Reviewed 4 of 13 files at r1, 4 of 10 files at r2, 2 of 255 files at r3, 253 of 253 files at r4.
Review status: 263 of 266 files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

marksvc commented:
Thanks for your work on this. Looks like the `<MajorVersion>` thing still needs addressed?

---

Reviewed 5 of 255 files at r3, 10 of 253 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[build/FLExBridge.proj, line 171 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyYhrB-NAql0rARqZs:-KzyYhrB-NAql0rARqZt:b-lqj6vb) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/build/FLExBridge.proj#L171)):*
> ```XML
> 			DateFormat="yyyy" DatePlaceholder="DEV_RELEASE_YEAR"
> 			Regex="DEV_VERSION_NUMBER" ReplacementText="$(Version)"/>
> 			<!-- ReplacementText is required, and must appear in the final file for date stamping to work. -->
> ```

What? I don't understand what this comment is saying.

---

*[build/WixPatchableInstaller.targets, line 34 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZmS1LR_a-m2na3GJP:-KzyVKfb-qP2BgvP8h5G:b-6cqj6) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L34)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

They're prettier in green. How would you feel about`<!-- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -->`?
</blockquote></details>

In general I'm not a fan of banner lines in source code. But I know others like them. You can put in a dashed line banner if you wish.

---

*[src/Installer/Installer.wxs, line 24 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyZCwEIBlGviedgCmj:-KzyZCwEIBlGviedgCmk:b-atyauv) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/src/Installer/Installer.wxs#L24)):*
> ```XML
> 
> 	<UIRef Id="WixUI_Minimal"/>
> 	<WixVariable Id="WixUILicenseRtf" Value="..\..\DistFiles\License.htm"/>
> ```

The Id is still `WixUILicenseRtf`. Is that an omission , or is that to help with upgrades?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-:-Kzyah7aLO8SNNGYgI7t:b-daxtnu)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/FLExBridge.proj, line 171 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyYhrB-NAql0rARqZs:-Kzz-6r4ADCgPdH5qK8u:bo26grw) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/build/FLExBridge.proj#L171)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

What? I don't understand what this comment is saying.
</blockquote></details>

I seem to have sacrificed clarity for brevity.

The FileUpdate task checks whether ReplacementText is in the updated file contents before writing them back to the file, but we need to replace two dates and only one Regex, so we call FileUpdate twice with the same Regex and ReplacementText.

I anticipate this will confuse future developers, although the  task will quickly fail if ReplacementText is altered. Do you think this deserves a comment, or shall we allow future developers who care to waste 5-15 minutes on this?

---

*[build/WixPatchableInstaller.targets, line 73 at r2](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzZnZA34r2D9LIt6egv:-KzzFLAFOdXQbJGQhdrL:b-rrv1qy) ([raw file](https://github.com/sillsdev/flexbridge/blob/92760b1ff627acb99fc6dd60d072398f36bfdd7d/build/WixPatchableInstaller.targets#L73)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

You're right; maybe I should think on this after Thanksgiving.
</blockquote></details>

well, now that it's after Thanksgiving, I suppose I can take a look.

---

*[src/Installer/Installer.wxs, line 24 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyZCwEIBlGviedgCmj:-Kzz55mrqC5yV5bPaCpf:b-nmgnb2) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/src/Installer/Installer.wxs#L24)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

The Id is still `WixUILicenseRtf`. Is that an omission , or is that to help with upgrades?
</blockquote></details>

Thanks for making me think about that. WixUILicenseRtf is a magic variable, and I'm pretty sure it needs to be a path to an RTF file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

marksvc commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/FLExBridge.proj, line 171 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyYhrB-NAql0rARqZs:-L-7mLe5eSax98e6gK3o:bhopc1c) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/build/FLExBridge.proj#L171)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

I seem to have sacrificed clarity for brevity.

The FileUpdate task checks whether ReplacementText is in the updated file contents before writing them back to the file, but we need to replace two dates and only one Regex, so we call FileUpdate twice with the same Regex and ReplacementText.

I anticipate this will confuse future developers, although the  task will quickly fail if ReplacementText is altered. Do you think this deserves a comment, or shall we allow future developers who care to waste 5-15 minutes on this?
</blockquote></details>

I don't think I understand enough of how this works, or how likely it will be that a developer will be modifying this file. Certainly if you think this will easily trip someone up, the future developer will be glad for a helpful comment. 

---

*[build/WixPatchableInstaller.targets, line 59 at r5](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-L-7mwz0eqbM567aeM08:-L-7mwz1RUwnu5mHmcHg:bij6ha2) ([raw file](https://github.com/sillsdev/flexbridge/blob/593a894f792e301b6fed3713cbd75dc3052c7c5f/build/WixPatchableInstaller.targets#L59)):*
> ```XML
> 
> 			<!-- Build Directories -->
> 			<AppBuildDir>$(InstallersBaseDir)/$(SafeApplicationName)_$(MinorVersion)_Build</AppBuildDir>
> ```

Haha, now that's tricky, that requesting "MinorVersion" gives you the major version and the minor version. I didn't think about that. It will hopefully be clear in the context since the lines are just a few above.

---

*[src/Installer/Installer.wxs, line 24 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyZCwEIBlGviedgCmj:-L-7nVZuWNYIeDV6r4JH:b-llbxs3) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/src/Installer/Installer.wxs#L24)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

Thanks for making me think about that. WixUILicenseRtf is a magic variable, and I'm pretty sure it needs to be a path to an RTF file.
</blockquote></details>

Okay, but how does that help us do away with the .rtf file? I guess I'll just have to live with it. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->

papeh commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build/WixPatchableInstaller.targets, line 59 at r5](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-L-7mwz0eqbM567aeM08:-L-7ofPULiBEB5U1M-bo:bo9jc2u) ([raw file](https://github.com/sillsdev/flexbridge/blob/593a894f792e301b6fed3713cbd75dc3052c7c5f/build/WixPatchableInstaller.targets#L59)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Haha, now that's tricky, that requesting "MinorVersion" gives you the major version and the minor version. I didn't think about that. It will hopefully be clear in the context since the lines are just a few above.
</blockquote></details>

If you have an epiphany regarding naming, we can update the genericinstaller-sample app, too

---

*[src/Installer/Installer.wxs, line 24 at r4](https://reviewable.io:443/reviews/sillsdev/flexbridge/170#-KzyZCwEIBlGviedgCmj:-L-7oy0kJVQne7dg4ttt:bmxq5rc) ([raw file](https://github.com/sillsdev/flexbridge/blob/4f8e2ce01c0117994796381d010d2b17c25ecdd6/src/Installer/Installer.wxs#L24)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Okay, but how does that help us do away with the .rtf file? I guess I'll just have to live with it. :)
</blockquote></details>

for now, at least. The genericinstaller uses an .htm file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/flexbridge/170)*
<!-- Sent from Reviewable.io -->
